### PR TITLE
✨(go/v4-alpha): add new golang plugin which uses the default golang base v3 and the new alpha kustomze/v2-alpha

### LINKED ISSUE #2547: Create go/v4 plugin with alpha1 level to address changes which are breaking ones for go/v3
### What do you want to happen?

The goal of this task is to:
- Create go/v4 plugin with alpha1 level to address changes which are breaking ones for go/v3

Also, you can check the policy: https://github.com/kubernetes-sigs/kubebuilder/blob/master/VERSIONING.md#introducing-changes-to-plugins.

**What is required to:** 

- Make a copy of https://github.com/kubernetes-sigs/kubebuilder/tree/master/pkg/plugins/golang/v3 to https://github.com/kubernetes-sigs/kubebuilder/tree/master/pkg/plugins/golang/v4
- see that we define the level / version of the plugin in: https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/plugins/golang/v3/plugin.go#L29-L31
- Then, rename the plugin. info to go/v4 + ensure that it will be alpha1 version 
- We also need to began to scaffold samples for the go/v4 as we do for go/v3 under the testdata, see: https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/testdata/generate.sh

Screenshot 2022-03-18 at 15 45 03

For go/v4 we do not need the sample v1beta1 since we will remove its support. See: https://github.com/kubernetes-sigs/kubebuilder/issues/2531

- By latest, we also will need the e2e tests for go/v4 see: https://github.com/kubernetes-sigs/kubebuilder/tree/master/test/e2e/v3

### Extra Labels

_No response_

### PR DESCRIPTION
**Pull request by NikhilSharmaWe**
### Motivation
- Allow users to begin to use the new kustomze eaiser by just calling kubebuilder init --plugin=go/v4-alpha
- Add initial version for go/v4-alpha so that in follow-ups we can also create a go base v4 alpha plugin and begin to address changes like remove the logic to allow create v2beta1 CRDs and webhook which was removed with k8s 1.22
- Also, seems time for we in a follow up deprecate the go/v2 plugin which was created to ensure backward compatibility with the legacy layout of Kubebuilder < 3.0.0 and no longer can be updated

Closes: #2547

### CODE REVIEW

**Comment by NikhilSharmaWe** [comment]
@camilamacedo86 
- Till now I have only added `pkg/plugins/golang/v4/` and changed plugin version.
https://github.com/kubernetes-sigs/kubebuilder/blob/fb7c9f303282ab6972a3c812d30a37bf024ba15e/pkg/plugins/golang/v4/plugin.go#L28-L31
Do we need to add something like `[]config.Version{config.Version{Number : 4, Stage: stage.Alpha}}` in `supportedProjectVersions` ? 
- Also about adding `pkg/config/v4`. Changes in the logic will be needed in `config.go` and `config_test.go` files for the purpose as I saw the code inside those files. Could you explain about the approach we should take.

**Comment by camilamacedo86** [comment]
Hi @NikhilSharmaWe,

> Do we need to add something like []config.Version{config.Version{Number : 4, Stage: stage.Alpha}} in supportedProjectVersions ?

Yes, we need to define its level as alpha: 
https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/model/stage/stage.go#L35-L36

Also, this PR is missing the testdata samples and the integration tests (e2e)


> Also about adding pkg/config/v4. Changes in the logic will be needed in config.go and config_test.go files for the purpose as I saw the code inside those files. Could you explain about the approach we should take.

The config defines the Project config which is the structure in https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v3/PROJECT. The go/v4 plugin can still be using the project 3 config for now. 
See: https://github.com/kubernetes-sigs/kubebuilder/blob/master/VERSIONING.md#versioning

c/c @Ankit152

**Comment by NikhilSharmaWe** [comment]
@camilamacedo86 
Could you please confirm
- **Under kubebuilder/testdata**
 Add files for v4 similar to the files for v3 except [project-v3-v1beta1](https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata/project-v3-v1beta1).
- **Under kubebuilder/test/e2e/v4**
Copy paste the files under kubebuilder/test/e2e/v3.

**Comment by NikhilSharmaWe** [comment]
@camilamacedo86 Made changes according to [this](https://github.com/kubernetes-sigs/kubebuilder/pull/2552#issuecomment-1073170689) comment.

**Comment by camilamacedo86** [comment]
@NikhilSharmaWe 

Not passing in the lint: https://github.com/kubernetes-sigs/kubebuilder/pull/2546#issuecomment-1070983260
To fix run make lint-fix in this case

c/c @Ankit152

File: docs/book/src/reference/watching-resources/testdata/external-indexed-field/api.go
Comment (camilamacedo86) [inline docs/book/src/reference/watching-resources/testdata/external-indexed-field/api.go:22]: Ideally, we should not touch on these files because they are not part of the scope.
However, I see that these changes are more related to format so that is fine

File: pkg/plugins/golang/v4/plugin.go
Comment (camilamacedo86) [inline pkg/plugins/golang/v4/plugin.go:21]: run. make lint-fix to fix this one.

File: test/e2e/v4/generate_test.go
Comment (camilamacedo86) [inline test/e2e/v4/generate_test.go:138]:
```suggestion
		"--plugins", "go/v4",
```

File: test/e2e/v4/generate_test.go
Comment (camilamacedo86) [inline test/e2e/v4/generate_test.go:133]:
```suggestion
func GenerateV4(kbc *utils.TestContext, crdAndWebhookVersion string) {
```

File: test/e2e/v4/generate_test.go
Comment (camilamacedo86) [inline test/e2e/v4/generate_test.go:134]: See that we should not copy all we need to add the test to generate and run the go/v4 instead

File: testdata/project-v4-addon/PROJECT
Comment (camilamacedo86) [inline testdata/project-v4-addon/PROJECT:3]: These samples were not built with the new alpha plugin

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
.

**Comment by NikhilSharmaWe** [comment]
Could you please review the changes made in 
- [test/e2e/v4/generate_test.go](https://github.com/kubernetes-sigs/kubebuilder/pull/2552/files/c9c52efb24960d2d86ce3a311f92b075f01d9f58#diff-a132e584478a89842ab0f0e505efd07b930759da970d10af6229f6944bd72992)
- https://github.com/kubernetes-sigs/kubebuilder/blob/c469a16b2f2345d5972d386163e92e22d2e07e7a/test/e2e/v4/plugin_cluster_test.go
- testdata/project-v4-addon/PROJECT

**Comment by Ankit152** [comment]
@NikhilSharmaWe I was going through the code and seems like you have not made any changes to the `test/testdata/generate.sh`. Kindly add the project commands there. I think that's needed to make the `go/v4` work.

Let me know your thought on this. 🙂 

c/c @camilamacedo86

**Comment by NikhilSharmaWe** [comment]
@camilamacedo86 

**Regarding changes in [test/testdata/generate.sh](https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/testdata/generate.sh) for scaffolding test projects**
- Should I add conditions similar to which are made for 
`project-v3`, `project-v3-multigroup`, `project-v3-addon`, `project-v3-config`
for 
`project-v4`, `project-v4-multigroup,` `project-v4-addon`, `project-v4-config`.
- Could you please respond to [this](https://github.com/kubernetes-sigs/kubebuilder/pull/2552#issuecomment-1073179136) comment.
- Also what others changes we need before we work on https://github.com/kubernetes-sigs/kubebuilder/issues/2531.

**Comment by camilamacedo86** [comment]
Hi @NikhilSharmaWe, 

> Regarding changes in [test/testdata/generate.sh](https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/testdata/generate.sh) for scaffolding test projects
Should I add conditions similar to which are made for
project-v3, project-v3-multigroup, project-v3-addon, project-v3-config
for
project-v4, project-v4-multigroup, project-v4-addon, project-v4-config. 

Yes, but all should be built with the go/v4 plugin so that is copied and past and then tweak. 

> Could you please respond to https://github.com/kubernetes-sigs/kubebuilder/pull/2552#issuecomment-1073179136 comment.

Done, already, see that I added comments on this PR. 
You need to change all to use the new go/v4 plugin

> Also what others changes we need before we work on https://github.com/kubernetes-sigs/kubebuilder/issues/2531.

When we have the go/v4 plugin in alpha (this PR merged) then is possible to work on this another task in a follow up

File: docs/book/src/reference/watching-resources/testdata/owned-resource/controller.go
Comment (camilamacedo86) [inline docs/book/src/reference/watching-resources/testdata/owned-resource/controller.go:137]: The changes n the docs will make it hard we do the review and are not part of the scope 
Can you revert, please?

File: test/e2e/v4/generate_test.go
Comment (camilamacedo86) [inline test/e2e/v4/generate_test.go:38]: Why we are generating a go/v2 here?

Comment (NikhilSharmaWe) [reply test/e2e/v4/generate_test.go:38]: So we have to 
**Regarding test/e2e/**
- Remove `test/e2e/v4` directory.
- Add a function `GenerateV4` in `test/e2e/v3/generate_test.go` similar to `GenerateV3` but with plugin : go/v4 and version : 4.
- Add Context similar to https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v3/plugin_cluster_test.go#L98-L129
but with `Root : "plugin go.kubebuilder.io/v4"` in `test/e2e/v3/plugin_cluster_test.go` and use func `GenerateV4` instead in its logic.

Could you inform I am understanding correct here.

**Comment by NikhilSharmaWe** [comment]
@camilamacedo86 
So we have to 
**Regarding test/e2e/**
- Remove `test/e2e/v4` directory.
- Add a function `GenerateV4` in `test/e2e/v3/generate_test.go` similar to `GenerateV3` but with plugin : go/v4 and version : 4.
- Add Context similar to https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v3/plugin_cluster_test.go#L98-L129
but with `Root : "plugin go.kubebuilder.io/v4"` in `test/e2e/v3/plugin_cluster_test.go` and use func `GenerateV4` instead in its logic.

Could you inform am I understanding correct here based on your previous comment ?

**Comment by NikhilSharmaWe** [comment]
@camilamacedo86 Could you please respond to my DMs on slack.

File: cmd/main.go
Comment (camilamacedo86) [inline cmd/main.go:63]:
```suggestion
		cli.WithDefaultPlugins(cfgv4.Version, gov4Bundle),
```

Comment (camilamacedo86) [reply cmd/main.go:63]: @NikhilSharmaWe ˆ

**Comment by NikhilSharmaWe** [comment]
At this https://github.com/NikhilSharmaWe/kubebuilder/blob/newPlugin/test/e2e/v2/plugin_cluster_test.go#L89 statement test is failing. Also logs shows that cause of the error is different from before `panic: runtime error: invalid memory address or nil pointer dereference` , https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kubebuilder/2552/pull-kubebuilder-test/1505995450979717120.

File: pkg/plugins/golang/v4/plugin.go
Comment (camilamacedo86) [inline pkg/plugins/golang/v4/plugin.go:31]: It shows the reason for the issue. 
The plugin version is : 4 alpha
The Project Version ( Project file config ) : would be the same of go/v3 plugin -> Plugin Config version 3

File: test/e2e/v3/generate_test.go
Comment (camilamacedo86) [inline test/e2e/v3/generate_test.go:239]: Then,. when we properly ensure that the go/v4 ( bundle plugin in the cli.go ) be == alpha we will need to change the key. IT will use alpha as well.

**Comment by AlmogBaku** [comment]
since the title will be used to generate release notes I suggest writing something more modest like ":sparkels: Initial base-code for go/v4 alpha plugin"

**Comment by NikhilSharmaWe** [comment]
@AlmogBaku @camilamacedo86 
Error says : `Error: failed to initialize project: unable to run pre-scaffold tasks of \"base.go.kubebuilder.io/v3\": go version 'go1.18' is incompatible because 'plugin requires go1.17 <= version < go1.18'`, but I am using go1.17 version. So, what could be the reason behind these logs?

File: pkg/plugins/golang/v4/init.go
Comment (AlmogBaku) [inline pkg/plugins/golang/v4/init.go:38]: Min version should be go1.18 (see #2561)

**Comment by AlmogBaku** [comment]
Following up on our Slack chat:

1. The e2e tests are using the `kubekins-e2e` image
2. The image configures to use go 1.18 here https://github.com/kubernetes/test-infra/blob/55ba9f6da3dab1e4df7698e553a7e68661ee0839/images/kubekins-e2e/variants.yaml#L25

What I don't understand, is why do you get this error at all... because you configured the `goVerMax` to `go2.0alpha1`...

File: pkg/plugins/golang/v4/init.go
Comment (AlmogBaku) [inline pkg/plugins/golang/v4/init.go:38]: Min version should be go1.18 (see #2561)

**Review by AlmogBaku — CHANGES_REQUESTED** [review]

File: pkg/plugins/golang/v3/init.go
Comment (NikhilSharmaWe) [inline pkg/plugins/golang/v3/init.go:39]: Should I revert this change, since the e2e tests were failing because of the issue in the tests repo.

Comment (NikhilSharmaWe) [reply pkg/plugins/golang/v3/init.go]: Reverted

File: pkg/plugins/golang/v4/init.go
Comment (camilamacedo86) [reply pkg/plugins/golang/v4/init.go:38]: This plugin uses go 1.17.
Then, if this PR goes first we can update it in your PR or in a follow since this one is alpha

**Comment by NikhilSharmaWe** [comment]
@camilamacedo86 Do we also need to update the controller-runtime and k8s dependencies for go/v4 ? Are these the changes because of which the PR is at hold right now.

**Comment by camilamacedo86** [comment]
Hi @NikhilSharmaWe,

We might need to re-think/re-discuss this one, see :
 
- the kustomize change actually is in the kustomize plugin : https://github.com/kubernetes-sigs/kubebuilder/pull/2583
- Then, the only changes that we have so far for go/v4-alpha would be: 

a) use the new version of test libs and change its scaffolds
b) remove the crd-version and webhook version options
c) make the kustomize/v2 the default plugin used by it

So, how can we add the go/v4-alpha and do not be overwhelmed to keep all plugins maintained? 
- Could we just do 1 sample scaffold until it is stable (project-v4-alpha only)? I think so 
- Could we re-use the v3 code for this one where it is equal and do not increase the number of duplicate files? If yes, what would be the pros/cons? 

Therefore, regards this one, I think we will need to:

a) get the kustomize pr merged first
b) change the plugin bundle to use it
c) change the samples generate sh script to gen only 1 project/sample
d) add the doc over this plugin

As follow up:

a) use the new version of test libs and change its scaffolds
b) remove the crd-version and webhook version options

Regarding your question > Do we also need to update the controller-runtime and k8s dependencies for go/v4 ? 
That is the kind of burn that we need to think about before moving forward. 


So, let's see if we can move with the kustomize first (I would suggest we hold a little this one).

**Comment by NikhilSharmaWe** [comment]
@camilamacedo86 what are the progress with the kustomize? and on what more changes we are looking for before we can merge this one?
Please inform me if I can help anyway.

**Comment by camilamacedo86** [comment]
HI @NikhilSharmaWe, 

Until we be able to discuss and move forward with or not:

the kustomize change actually is in the kustomize plugin : https://github.com/kubernetes-sigs/kubebuilder/pull/2583 

Then, I do not think that we can discuss this one. ( So, the pre-requirement for we can back to lock here is to check the above PR merged OR closed ) Then, that means we will have the final required decisions to know what will be the best approach to move forward.

File: test/e2e/v3/generate_test.go
Comment (camilamacedo86) [inline test/e2e/v3/generate_test.go:408]: The v4 tests myd be in test/e2e/v4 instead 
Could you please ensure that we have a new dir with them?

**Review by camilamacedo86 — APPROVED** [review]
Thank you for the contribution 🥇 

IHMO we need to move forward here and add the go/v4-alpha to make it easier for users to begin to experiment kustomize v4 with and we begin to support Apple Silicon.

Also, it seems time to deprecate go/v2 since we no longer can update it and it was created only to ensure the backwards compatibility,

File: cmd/main.go
Comment (camilamacedo86) [inline cmd/main.go:45]:
```suggestion
	// Bundle plugin which built the golang projects scaffold by Kubebuilder go/v3 with kustomize alpha-v2
```
nit ^

**Review by everettraven — APPROVED** [review]
/lgtm

**Review by AlmogBaku — APPROVED** [review]

**Comment by camilamacedo86** [comment]
/hold cancel

we are good to move with this one :-)

**Pull request merged**
